core/src/data_sources/splitter.rs

```diff
@@ -1,13 +1,379 @@
+                            content: Some(t),
+                            sections: vec![],
+                        })
+                        .collect::<Vec<_>>(),
+                );
+
+                // Remove the content from the current section.
+                content = None;
+            }
+        }
+
+        // Create a new leaf children for content if prefix is not None and content is not None to
+        // preserve the invariant that content nodes are leaf nodes.
+        if let Some(c) = content.as_ref() {
+            if let Some(_) = prefix.as_ref() {
+                sections.push(TokenizedSection {
+                    max_chunk_size,
+                    prefixes: prefixes.clone(),
+                    tokens_count: prefixes_tokens_count + c.tokens.len(),
+                    content: Some(c.clone()),
+                    sections: vec![],
+                });
+
+                // Remove the content from the current section.
+                content = None;
+            }
+        }
+
+        sections.extend(
+            futures::future::join_all(
+                section
+                    .sections
+                    .iter()
+                    .map(|s| TokenizedSection::from(embedder, max_chunk_size, prefixes.clone(), s)),
+            )
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?,
+        );
+
+        Ok(TokenizedSection {
+            max_chunk_size,
+            prefixes,
+            // The tokens_count is `prefixes_tokens_count` to which we add:
+            // - the sum of the content tokens
+            // OR (or because content nodes are leaf nodes)
+            // - the tokens_count of the childrens to which we remove childrens times the
+            // `prefixes_tokens_count` as they would not be repeated in children if we were to
+            // reconstruct that node as a full chunk.
+            tokens_count: prefixes_tokens_count
+                + match content.as_ref() {
+                    Some(c) => c.tokens.len(),
+                    None => {
+                        sections.iter().map(|s| s.tokens_count).sum::<usize>()
+                            - sections.len() * prefixes_tokens_count
+                    }
+                },
+            content,
+            sections,
+        })
+    }
+
+    /// DFS traversal of the TokenizedSection tree to generate a vector.
+    fn dfs(&self) -> Vec<&Self> {
+        let mut res = vec![self];
+        for section in &self.sections {
+            res.extend(section.dfs());
+        }
+        res
+    }
+
+    /// This function materialize a chunk from a TokenizedSection node.
+    fn chunk(&self) -> TokenizedText {
+        let mut seen_prefixes: HashSet<String> = HashSet::new();
+        let mut text = String::new();
+        let mut tokens: Vec<usize> = vec![];
+
+        for s in self.dfs() {
+            s.prefixes.iter().for_each(|p| {
```

Comment:
Now we could do all the mother section prefixes, then just the last prefix for each section (no seen_prefixes logic)
A bit simpler IMHO but not sure it's worth the change